java/arcs/core/util/Scheduler.kt

```diff
@@ -0,0 +1,242 @@
+    }
+
+    private fun CoroutineScope.startProcessingJob() {
+        isIdle.value = false
+        processingJob = launch {
+            launches.incrementAndGet()
+
+            var shallContinue = true
+            while (shallContinue) {
+                val startTime = time.currentTimeMillis
+                shallContinue = withTimeout(agendaProcessingTimeout) { process() }
+                val elapsed = time.currentTimeMillis - startTime
+
+                if (shallContinue) {
+                    loops.incrementAndGet()
+
+                    // If the operation was super fast, let's delay a little bit to allow the
+                    // thread to chill out.
+                    val delayLength = scheduleRateHz.hzToMillisPerIteration() - elapsed
+                    if (delayLength > 0) delay(delayLength)
+                }
+            }
+        }.also { it.invokeOnCompletion { isIdle.value = true } }
+    }
+
+    private suspend fun process(): Boolean {
+        val timeoutHandler = { throwable: Throwable ->
+            if (throwable is TimeoutCancellationException) {
+                log.error(throwable) { "Scheduled tasks timed out." }
+            }
+        }
+
+        if (isPaused.value) return false
+
+        val agenda = agenda.getAndSet(Agenda())
+        if (agenda.isEmpty()) return false
+
+        log.debug { "Processing agenda: $agenda" }
+
+        // Process agenda
+        agenda.forEach { task ->
+            suspendCancellableCoroutine<Unit> { it.resume(task(), timeoutHandler) }
+        }
+
+        return true
+    }
+
+    private fun Int.hzToMillisPerIteration(): Long = ((1.0 / toDouble()) * 1000L).toLong()
+
+    sealed class Task(private val block: () -> Unit) {
+        /**
+         * [Task]s will be invoked by the [Scheduler] when their turn has arrived.
+         */
+        operator fun invoke() = block()
+
+        /**
+         * A [Processor] [Task] is responsible for computing data ahead of the execution of any
+         * [Listener] tasks.
+         *
+         * For example: [StorageProxy]'s usage of the [Scheduler] schedules [Processor] [Task]s when
+         * a [ProxyMessage] comes in from the [Store] - these [Processor] tasks are responsible for
+         * updating the [StorageProxy]'s local [CrdtModel].
+         */
+        abstract class Processor(block: () -> Unit) : Task(block)
+
+        /**
+         * [Listener] [Task]s are invoked after all scheduled [Processor] tasks have finished. The
+         * idea is that they are meant to react to changes that [Processor] tasks have triggered.
+         */
+        abstract class Listener(
+            /**
+             * In situations where [name]s can collide, it's helpful to have an additional dimension.
+             *
+             * For example: [StorageProxy]'s usage of [Scheduler] uses a handle's owning Particle's
+             * name as the [namespace].
+             */
+            val namespace: String,
```

Comment:
I'm curious why this is needed.  Since each ArcId gets its own HandleManager, and since each HandleManager has its own separate pool of proxies, would't they all just share the same Scheduler instance, but different Arcs or HandleManager instances would have a separate list, so aren't they protected from collisions? 


Also, in the case of two Particles with the same name and same handle name, would there be a problem? This is a recent bug we ran into. Particle names aren't unique, neither are ParticleName/HandleName pairs.

Reply:
Particle names aren't even unique per arc?

Reply:
Nope. You can include the same particle twice. This probably doesn't impact your PR, it just means we will use Particle instance id instead of name.